Formal2024/Workshops/workshop4.lean

```diff
@@ -0,0 +1,251 @@
+#check Type -- Lean has a countably infinite hierarchy of type of types (of types of types...)
+#check Type 1
+--#check Type 33 -- Well, almost infinite
+-- unless you manually increase the `maxUniverseOffset`
+set_option maxUniverseOffset 33
+#check Type 33
+
+#check Prop → ℕ -- We can construct types of functions between types
+#check Type 1 → Type 2
+
+-- There are two inference rules that tell us that functions work like we would expect
+
+variable {α β : Type*} -- With the asterisk we are leaving the level of the type generic
+#check α
+#check β
+variable (f : α → β)
+variable (a : α)
+
+#check f a -- We can apply functions, getting the expected outcome
+
+def myAnd (P : Prop) : Prop := P ∧ P
+
+#check fun (P: Prop) ↦ myAnd P -- We can perform abstractions to create functions
+
+
+-- Apart from `∀` (which we discussed is equivalent to functions) everything else is constructed
+-- starting from here (mostly using Inductive Types, as we will see).
+end
+
+
+section
+/-
+  # How is a type different from a set?
+-/
+
+def setNat : Set ℕ := univ -- This is the set of all natural numbers
+
+-- Sets come with a predicate: `∈` (`\in` or `\mem`). Its negation is `∉` (`\notin`)
+
+#check ( 1 ∈ setNat ) -- This is in `Prop`
+#check ( 1 : ℕ )     -- This is not
+
+-- `( 1 : ℕ )` is a statement in type theory, but not an element in `Prop`, our model of
+-- propositional logic inside type theory.
+
+-- Another big difference: if `a : α` and `a : β`, then `α` and `β` are definitionally equal
+-- To make sense of things like subsets, intersections, union, complements, we need `Set`
+
+end
+
+section
+variable {α : Type*}
+variable (s t u : Set α)
+
+-- We write `⊆` (`\ss` or `\sub`) for subset, `∩` (`\i` or `\cap`) for intersection,
+-- `∪` (`\u` or `\cup`) for union, and `\` (typed as `\`) for set difference.
+
+-- We can also use the notation `{ y | P y }`, where `P` is a predicate, for the set of
+-- elements satisfying `P`
+
+#check subset_def
+#check mem_setOf
+#check inter_def
+#check union_def
+#check mem_diff
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  simp only [subset_def, inter_def, inter_def, mem_setOf]
+  rw [subset_def] at h
+  -- We can use `simp only` or `rw` to unfold the definitions
+  -- `simp only` is a more powerful `rw`, that also works under quantifiers
+  -- Now we can proceed as usual
+  rintro x ⟨xs, xu⟩
+  exact ⟨h x xs, xu⟩
+  tada
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  -- This is not needed though, as the anonymus constructor forces Lean to unfold the definitions
+  rintro x ⟨xs, xu⟩
+  exact ⟨h xs, xu⟩ -- There may be some slight differences (here we do not need to specify x)
+  tada
+
+
+example : s ∩ (t ∪ u) ⊆ s ∩ t ∪ s ∩ u := by
+  intro x hx                      -- Subset works like an implication
+  have xs : x ∈ s := hx.1         -- Intersection works like an `∧`
+  have xtu : x ∈ t ∪ u := hx.2    -- Union works like an `∨`
+  rcases xtu with xt | xu
+  · left
+    show x ∈ s ∩ t
```

Comment:
I believe this is the first time we've used `show`. We should probably say something about it.

Reply:
To be fair here show really does nothing. I changed it to unfolding the definition of intersection and added the comment.